internal/validator/validator.go

```diff
@@ -0,0 +1,317 @@
+// applications.
+type Validator struct {
+	repository              ValidatorRepository
+	inputBoxDeploymentBlock uint64
+}
+
+func NewValidator(
+	repo ValidatorRepository,
+	inputBoxDeploymentBlock uint64,
+) *Validator {
+	return &Validator{repo, inputBoxDeploymentBlock}
+}
+
+func (v *Validator) String() string {
+	return "validator"
+}
+
+// Run executes the Validator main logic of producing claims and/or proofs
+// for processed epochs of all running applications. It is meant to be executed
+// inside a loop. If an error occurs while processing any epoch, it halts and
+// returns the error.
+func (v *Validator) Run(ctx context.Context) error {
+	apps, err := v.repository.GetAllRunningApplications(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get running applications. %w", err)
+	}
+
+	for idx := range apps {
+		if err := v.validateApplication(ctx, apps[idx]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validateApplication calculates, validates and stores the claim and/or proofs
+// for each processed epoch of the application.
+func (v *Validator) validateApplication(ctx context.Context, app Application) error {
+	slog.Info("starting validation", "application", app.ContractAddress, "service", v)
+	processedEpochs, err := v.repository.GetProcessedEpochs(ctx, app.ContractAddress)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get processed epochs of application %v. %w",
+			app.ContractAddress, err,
+		)
+	}
+
+	for _, epoch := range processedEpochs {
+		slog.Info("started calculating claim",
+			"epoch index", epoch.Index,
+			"application", app.ContractAddress,
+			"service", v,
+		)
+		claim, outputs, err := v.createClaimAndProofs(ctx, epoch)
+		slog.Info("finished calculating claim",
+			"epoch index", epoch.Index,
+			"application", app.ContractAddress,
+			"service", v,
+		)
+		if err != nil {
+			return err
+		}
+
+		// The Cartesi Machine calculates the root hash of the outputs Merkle
+		// tree after each input. Therefore, the root hash calculated after the
+		// last input in the epoch must match the claim hash calculated by the
+		// Validator. We first retrieve the hash calculated by the
+		// Cartesi Machine...
+		machineClaim, err := v.repository.GetLastInputOutputHash(
+			ctx,
+			epoch.Index,
+			epoch.AppAddress,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to get the machine claim for epoch %v of application %v. %w",
+				epoch.Index, epoch.AppAddress, err,
+			)
+		}
+
+		// ...and compare it to the hash calculated by the Validator
+		if machineClaim != nil && *machineClaim != *claim {
+			return fmt.Errorf(
+				"validator claim does not match machine claim for epoch %v of application %v",
+				epoch.Index, epoch.AppAddress,
+			)
+		}
+
+		// update the epoch status and its claim
+		epoch.Status = EpochStatusClaimComputed
+		epoch.ClaimHash = claim
+
+		// store the epoch and proofs in the database
+		if err := v.repository.SetEpochClaimAndInsertProofsTransaction(
```

Comment:
I prefer to no use this syntax if it means breaking multiple lines.
Do `err = ...` in the previous line, than do `if err != nil ...`.

Reply:
Done